### PR TITLE
(WIP) Limit map pins

### PR DESCRIPTION
**Pull request by stoe**
## :construction: WIP :construction:

**:exclamation: Do NOT merge :exclamation:**

As described in https://github.com/github/services-web/issues/655#issuecomment-276651827, this PR adds limiting the number of pins returned in `createMap.topojson`.

Merging now won't have an effect right now as our pin files don't provide a `date` property needed for the sorting to work.

### CODE REVIEW

File: _layouts/renderMap.topojson [hunk 1]
```diff
@@ -1,46 +1,22 @@
...
-  ]
-  }
-
-{% if page.showJS %}
-)
-{% endif %}
+{% assign sorted_pins = (site.pins | sort: 'date') | reverse %}
+{% if page.showJS %}eqfeed_callback({% endif %}{
+  "type": "FeatureCollection",
+  "features": [{% for pin in sorted_pins limit:100 %}
```
Comment (swinton) [inline _layouts/renderMap.topojson]: We might want to increase the limit here, the map may look a little sparse with only 100 pins?

**Comment by swinton** [comment]
Looks good to me, verified locally with `jq`:

```bash
curl http://127.0.0.1:4000/createMap.topojson | jq '.["features"] | length'
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 29570  100 29570    0     0  12.3M      0 --:--:-- --:--:-- --:--:-- 14.1M
100
```

:tada:

Looks like we have some bad data though, I added 3500ef3 to handle cases where we don't have a lat, lon pair. I'm not sure how these will be handled by the JavaScript, we might want to enforce this with CI?

**Comment by crichID** [comment]
> Looks like we have some bad data though, I added 3500ef3 to handle cases where we don't have a lat, lon pair.

@stoe @swinton I believe this is something we are already testing? Did something slip through the cracks?

**Comment by swinton** [comment]
> Did something slip through the cracks?

Here's one that's giving us a problem:

https://github.com/githubschool/open-enrollment-classes-introduction-to-github/blob/master/_pins/thc4agithub.json

Apparently due to a missing `---` at the foot of the file, when this is missing the data is passed through to [the template](https://github.com/githubschool/open-enrollment-classes-introduction-to-github/blob/master/_layouts/renderMap.topojson).

There are a few more like this:

- https://github.com/githubschool/open-enrollment-classes-introduction-to-github/blob/master/_pins/Madhura-a-kulkarni.json
- https://github.com/githubschool/open-enrollment-classes-introduction-to-github/blob/master/_pins/Milliye.json
- https://github.com/githubschool/open-enrollment-classes-introduction-to-github/blob/master/_pins/Shiv.json
- https://github.com/githubschool/open-enrollment-classes-introduction-to-github/blob/master/_pins/dabouelhassan.json
- https://github.com/githubschool/open-enrollment-classes-introduction-to-github/blob/master/_pins/denis-adobe.json
- https://github.com/githubschool/open-enrollment-classes-introduction-to-github/blob/master/_pins/devin-hanaway.json
- https://github.com/githubschool/open-enrollment-classes-introduction-to-github/blob/master/_pins/dianavazquez.json
- https://github.com/githubschool/open-enrollment-classes-introduction-to-github/blob/master/_pins/halimat94-london.json
- https://github.com/githubschool/open-enrollment-classes-introduction-to-github/blob/master/_pins/iharGoodman.json
- https://github.com/githubschool/open-enrollment-classes-introduction-to-github/blob/master/_pins/ijmarquez.json
- https://github.com/githubschool/open-enrollment-classes-introduction-to-github/blob/master/_pins/jaj371.json
- https://github.com/githubschool/open-enrollment-classes-introduction-to-github/blob/master/_pins/leti-ulloa.json
- https://github.com/githubschool/open-enrollment-classes-introduction-to-github/blob/master/_pins/thc4agithub.json
- https://github.com/githubschool/open-enrollment-classes-introduction-to-github/blob/master/_pins/way2desispirit.json

**Comment by swinton** [comment]
> Apparently due to a missing --- at the foot of the file

### Next steps

1. Add a test for this, to ensure the `---` file endings are present
1. Retroactively fix [the above files](https://github.com/githubschool/open-enrollment-classes-introduction-to-github/pull/3340#issuecomment-277053698)
1. Retract 3500ef3

**Comment by swinton** [comment]
> Add a test for this, to ensure the --- file endings are present
> Retroactively fix the above files

Happening in #3429.

**Comment by hectorsector** [comment]
@T4selogistics please refrain from closing issues and PRs owned by other folks. This one in particular, is ongoing work attempting to limit the map pins for the activity.

**Comment by hollenberry** [comment]
@easpaas please refrain from closing issues and PRs owned by other folks. This one in particular, is ongoing work attempting to limit the map pins for the activity.

**Comment by stoe** [comment]
Closing this as stale.
See also https://github.com/githubschool/open-enrollment-classes-introduction-to-github/pull/5146

**Comment by hectorsector** [comment]
@stoe I want to pick up where you left off. Could you help me with a tl;dr of what needs to be done v. where we are?

**Comment by stoe** [comment]
**TL;DR**:
We need to limit the number of pins.

We can achiev this in a few different ways
- In JavaScript `mapAPI.js`, not optimal as we wouls still load "all files" before limiting (not implemented)
- In Jekyll (this PR)
  1. problem: Files requested in `renderMap.topojson` are not sorted by default, sorting with Jekyll is slow
  2. problem: Files number is not limited and so it needs to be done prgrammatically in `renderMap.topojson`, which is also slow

Generally it can be done with the code in this PR @hectorsector, but I think we just never finished it.

**Comment by swinton** [comment]
I think we also discussed a script that would archive old pins, that we
would run manually periodically or via a cron.

**Comment by stoe** [comment]
Thanks, @swinton, you're right

**Comment by swinton** [comment]
I’m thinking it could be a bash script in /script that read the pins and moves old pins into a separate folder, archiving them so that we maintain a copy of the data but they don’t render on the map.

**Comment by stoe** [comment]
> I’m thinking it could be a bash script in /script that read the pins and moves old pins into a separate folder, archiving them so that we maintain a copy of the data but they don’t render on the map.

Sure, but that still would require running it manually in a defined cadence.

**Comment by stoe** [comment]
@swinton has a point.
As a MVP we can use a script like that for manual housekeeping when we realise the pins atart to render slow.

**Comment by swinton** [comment]
> As a MVP we can use a script like that for manual housekeeping when we realise the pins atart to render slow.

Yep, @stoe :100:

**Comment by hectorsector** [comment]
> I’m thinking it could be a bash script in /script that read the pins and moves old pins into a separate folder, archiving them so that we maintain a copy of the data but they don’t render on the map.

This would work! @githubschool/training-team already runs a few manual scripts for cleanup and we can easily add this to the bunch.

**Comment by stoe** [comment]
`find ./_pins -mtime +180 -type f -exec git mv '{}' ./_pins_archive \;` could work.

`./_pins` directory
`-mtime +180` modified older than 180 days, alternate `ctime` change time or `atime` last access time
`-type f` only files
`-exec git mv '{}' ./_pins_archive \;` execute the mv command

Problem is, when executed locally, any of the times are when the file was pulled (and thus created), so it is not acurate, or won't even work if I just cloned the repo.

@hectorsector @swinton for 👀 and 💭

**Comment by swinton** [comment]
Can we query `Git` for the date / time a file was first committed?

**Comment by brianamarie** [comment]
> Can we query `Git` for the date / time a file was first committed?

@swinton Yep, sure can: 

`git log --diff-filter=A -- foo.js`

**Comment by swinton** [comment]
> Problem is, when executed locally, any of the times are when the file was pulled (and thus created), so it is not acurate, or won't even work if I just cloned the repo.

Perhaps we can rely of [Git for the date times](https://github.com/githubschool/open-enrollment-classes-introduction-to-github/pull/3340#issuecomment-343476254) instead of the filesystem, @stoe?

**Comment by stoe** [comment]
`git ls-tree --name-only -r < commit-ish > -- ./_pins` could be faster, @brianamarie, but we'de need the `<commit-ish>` / SHA, which we probably can get via another command.

source: https://stackoverflow.com/a/444317

**Comment by hectorsector** [comment]
I played with this a little bit, and this could work. This is what my script is looking like:

```shell
# pick a date, this will probably be an argument when executing the script
chosenDate=$(date -j -v-365d +"%Y-%m-%d")

# determine the SHA1 closest to this date
commitid=$(git rev-list -1 --before=chosenDate master)

# checkout to that commit, move all the pins
git checkout $commitid
mv _pins/* backups

# create a new branch and commit the renames 
git checkout -b archive-pins master
git add _pins backups
git commit -m "backup pins prior to $chosenDate"

# ask user to verify everything looks right, and push + PR or just merge into master
```

**Comment by stoe** [comment]
`git rev-list -1 --before=@{1.year.ago} master` might be better, @hectorsector.
Git is quite smart when it comes to dates. 

The script could then be:

```sh
# determine the SHA1 closest to this date
commitid=$(git rev-list -1 --before=@{1.year.ago} master)

# checkout to that commit, move all the pins
git checkout $commitid
mv _pins/* backups

# create a new branch and commit the renames 
git checkout -b archive-pins master
git add _pins backups
git commit -m "backup pins prior to $chosenDate"

# ask user to verify everything looks right, and push + PR or just merge into master
```

**Comment by stoe** [comment]
`mv _pins/* backups` will fail if the `backup/` directory does not exist prior to the line being executed, @hectorsector.

What you think we replace it with `find ./_pins -type f -exec mv '{}' ./backups \;`?

**Comment by hectorsector** [comment]
😮 thanks for cleaning it up, @stoe! Judging by #11863 looks like it cleaned up nicely.
